design/fast_finality.md

```diff
@@ -0,0 +1,190 @@
+
+### Discovery of rollups
+
+Nodes scan incoming L1 blocks for new rollups. They validate each new rollup by:
+
+* Checking that it is produced by the designated sequencer, based on the sequencer listed in the management contract
+* Checking that it contains all the light batches produced since the last rollup. Each light batch contains the number 
+  of the rollup that will contain it. Since the rollup is a sparse Merkle tree, proving non-inclusion of a given light 
+  batch is straightforward
+
+They then persist the rollup, so that they have a record of which light batches have been confirmed on the L1.
+
+## Future work
+
+* Allowing nodes to challenge the sequencer's rollups (e.g. if the light batches are missing transactions, or if a 
+  certain light batch is not included in the rollup)
+* Creation of an inbox to allow transactions to be "forced through" if the sequencer is excluding them
+* High-availability of the sequencer
+
+## Unresolved issues
+
+* Select a design for preventing value extraction (see the section 
+  `Designs considering for preventing value-extraction`, below)
+* How do we ensure the sequencer distributes all light batches to all nodes?
+* Do the light batches need to be linked to the latest block that was fed into the enclave?
+* How do we achieve high-availability of the sequencer, both for the enclave and for the host? How do we prevent 
+  denial-of-service attacks on both?
+* How do we prevent attackers from "gumming up" the sequencer with random, fake light-batches?
+
+## Appendices
+
+### Designs considering for preventing value-extraction
+
+In this section, we investigate various designs to prevent value-extraction. The specific attack we have in mind is one 
+where the sequencer runs `n` enclaves, and uses `n-1` of them to test the impact of various transaction sets to 
+identify value-extraction opportunities.
+
+#### Do nothing
+
+In this approach, we rely on trust in the sequencer and the fact that value-extraction opportunities are reduced in
+Obscuro due to its data-visibility rules.
+
+#### Disable `eth_call` on sequencer enclaves
+
+By disabling `eth_call` on the sequencer enclaves, we prevent the operator from extracting information about the impact
+of a given light batch.
+
+This approach is unworkable. The operator can run a separate, non-sequencer enclave, feed the light batches to that
+enclave, and use `eth_call`s on that enclave to determine the impact of a given light batch.
+
+#### Run a single sequencer enclave
+
+If there is a single sequencer enclave, there are no other enclaves to use to identify value-extraction opportunities. 
+The single sequencer cannot be restarted to identify value-extraction opportunities, since the enclave start-up delay 
+will then prevent the sequencer from reaching its block production target.
+
+This approach is unworkable. We cannot achieve the desired high-availability with a single sequencer enclave.
```

Comment:
I don't think we should call this unworkable yet. Maybe we can achieve HA through the sequencer pool and not allow for HA per operator. (With initial centralised solution being a whitelisted pool).

Reply:
To expand on this, I think we should consider whether HA is a requirement too. HA isn't 100%, and I don't know if we buy enough uptime to be worth the complexity.

Reply:
Re: your first point, I think the creation of a sequencer pool is not a solution in itself - the enclaves of the other sequencers in the pool could be doing front-running, for example.

Re: your second point - true, I updated the requirements.

Reply:
Yeah agreed. But it changes what we're trying to solve from front-running in HA to front-running in sequencer re-election. And we'll need to solve that second one anyway when we get to decentralising.

Reply:
I've reworded it a bit, but I still feel like running a sequencer pool is by definition not running a single sequencer enclave.